### PR TITLE
[core] Support generators to allow tasks to return a dynamic number of objects

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

This adds support for tasks that need to return a dynamic number of objects. When a remote generator function is invoked and `num_returns` for the task is 1, the worker will dynamically allocate `ray.put` IDs for these objects and store an `ObjectRefGenerator` as its return value. This allows the worker to choose how many objects to return and to keep heap memory low, since it does not need to keep all objects in memory simultaneously.

Unlike normal `ray.put()`, we assign the task caller as the owner of the object. This is to improve fault tolerance, as the owner can recover dynamically generated objects through the normal lineage reconstruction codepath.

The main complication has to do with notifying the task caller that it owns these objects. We do this in two places, which is necessary because the protocols are asynchronous, so either message can arrive first.
1. When the task reply is received.
2. When the primary raylet subscribes to the eviction notice from the owner.
To register the dynamic return, the owner adds the ObjectRef to the ref counter and marks that it is contained in the generator object.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

TODO:
- docs

### CODE REVIEW

**Comment by stephanie-wang** [comment]
Still have some TODOs, but the main changes are ready for review.

**Review by clarkzinzow — COMMENTED** [review]
Sweet, general approach LGTM but I'll defer to others for a more thorough review.

In addition to the object eviction subscription, is there a potential race between the pinned object location update and the push task reply for the dynamic return objects? IIRC both of those RPCs are async with no synchronization barrier in-between. Maybe this is a non-issue because we always add the pinned object location when processing the push task reply? https://github.com/ray-project/ray/blob/775ff3e7bd818a7a1949dd0dd0a4ac4195de0f99/src/ray/core_worker/task_manager.cc#L249-L252

**Comment by stephanie-wang** [comment]
> Sweet, general approach LGTM but I'll defer to others for a more thorough review.
> 
> In addition to the object eviction subscription, is there a potential race between the pinned object location update and the push task reply for the dynamic return objects? IIRC both of those RPCs are async with no synchronization barrier in-between. Maybe this is a non-issue because we always add the pinned object location when processing the push task reply?
> 
> https://github.com/ray-project/ray/blob/775ff3e7bd818a7a1949dd0dd0a4ac4195de0f99/src/ray/core_worker/task_manager.cc#L249-L252

Hmm I think this part is okay because of what you said: it all happens locally at the task caller when processing the push task reply.

I think there may be a race condition for object directory location updates, though, let me look into this.

**Comment by stephanie-wang** [comment]
> > Sweet, general approach LGTM but I'll defer to others for a more thorough review.
> > In addition to the object eviction subscription, is there a potential race between the pinned object location update and the push task reply for the dynamic return objects? IIRC both of those RPCs are async with no synchronization barrier in-between. Maybe this is a non-issue because we always add the pinned object location when processing the push task reply?
> > https://github.com/ray-project/ray/blob/775ff3e7bd818a7a1949dd0dd0a4ac4195de0f99/src/ray/core_worker/task_manager.cc#L249-L252
> 
> Hmm I think this part is okay because of what you said: it all happens locally at the task caller when processing the push task reply.
> 
> I think there may be a race condition for object directory location updates, though, let me look into this.

Okay, I believe I've resolved this issue by also attaching the generator ID to spill location updates. I don't think it's necessary for in-memory locations (see added comments).

File: python/ray/_raylet.pyx
Comment (ericl) [inline python/ray/_raylet.pyx:168]:
```suggestion
        self._refs = refs
```

**Review by ericl — APPROVED** [review]
Just tried this, very cool! A couple API comments:
- When you specify num_returns, it seems that you can also yield the exact same number of values, but it is not a generator return. This seems confusing, should we disallow mixing num_returns and generators?
- Should we define `__len__` on the generator return object? It seems not unreasonable to include this, and even if we decide in the future to support streaming generators, we could just raise an error trying to get the length in that case.

**Review by ericl — COMMENTED** [review]
We should also add docs to this I think.

File: python/ray/tests/test_generators.py
Comment (jjyao) [inline python/ray/tests/test_generators.py:213]: Can generator itself be passed around as task args?

Can generator_obj_ref be passed around like other obj refs?

**Review by jjyao — COMMENTED** [review]
Will continue tomorrow.

File: python/ray/_raylet.pyx
Comment (stephanie-wang) [reply python/ray/_raylet.pyx:812]: Oh it might be fixed already... let me check :D

File: python/ray/_raylet.pyx
Comment (stephanie-wang) [reply python/ray/_raylet.pyx:862]: It happens during reconstruction, will add a comment.

File: python/ray/tests/test_generators.py
Comment (stephanie-wang) [reply python/ray/tests/test_generators.py:213]: Yes, and yes. I'll add test cases for these.

**Comment by ericl** [comment]
Yeah, that sounds pretty good. To make sure, you're proposing Alternative (2) below right?

```
Current PR:
-----------
plain return, num_returns=K -> returns Tuple_K[ObjectRef]
plain return, num_returns=None -> returns ObjectRef
yield, num_returns=K -> returns Tuple_K[ObjectRef]
yield, num_returns=None -> returns ObjectRef[RefGenerator]

Alternative 1:
--------------
plain return, num_returns=K -> returns Tuple_K[ObjectRef]
plain return, num_returns=None -> returns ObjectRef
yield, num_returns=K -> disallowed
yield, num_returns=None -> returns ObjectRef[RefGenerator]

Alternative 2:
--------------
plain return, num_returns=K -> returns Tuple_K[ObjectRef]
plain return, num_returns=None -> returns ObjectRef
yield, num_returns=K -> returns Tuple_K[ObjectRef]
yield, num_returns=None -> returns ObjectRef
yield, num_returns="dynamic" -> returns ObjectRef[RefGenerator]
```

File: python/ray/tests/test_generators.py
Comment (stephanie-wang) [reply python/ray/tests/test_generators.py:95]: Ah yes, will do!

File: python/ray/_raylet.pyx
Comment (stephanie-wang) [reply python/ray/_raylet.pyx:2188]: Ah nice catch, this was leftover from debugging...

File: python/ray/_raylet.pyx
Comment (stephanie-wang) [reply python/ray/_raylet.pyx:2163]: It gets passed to `store_task_output` so that we know who the owner is for dynamic objects.

File: src/ray/core_worker/core_worker.cc
Comment (stephanie-wang) [reply src/ray/core_worker/core_worker.cc:2439]: At this point, we don't know how many objects will be returned, since this happens one at a time for each yield statement. I prefer to use `GetNextPutIndex` because it already has the semantics of incrementing a counter for each new object.

That reminds me, though, let me add a test for recovery where the task also calls `ray.put`.

File: src/ray/raylet/local_object_manager.h
Comment (scv119) [inline src/ray/raylet/local_object_manager.h:240]: nits: comment need update.

File: src/ray/raylet/local_object_manager.h
Comment (scv119) [inline src/ray/raylet/local_object_manager.h:85]: nits: comments need update.

File: python/ray/tests/test_generators.py
Comment (scv119) [inline python/ray/tests/test_generators.py:219]: also wondering the behavior when generator failed after yield a few objects. would retry ever work in that case?

File: src/ray/protobuf/core_worker.proto
Comment (scv119) [inline src/ray/protobuf/core_worker.proto:220]: that this object should already be..?

File: src/ray/object_manager/ownership_based_object_directory.cc
Comment (scv119) [inline src/ray/object_manager/ownership_based_object_directory.cc:191]: hmm looks it's not used?

File: python/ray/_raylet.pyx
Comment (jjyao) [reply python/ray/_raylet.pyx]: Do we have test case that can catch this in the future?

File: doc/source/ray-core/tasks/generators.rst
Comment (jjyao) [inline doc/source/ray-core/tasks/generators.rst:19]: Could we explain why: why static generator is preferred where possible?

File: doc/source/ray-core/tasks/generators.rst
Comment (jjyao) [inline doc/source/ray-core/tasks/generators.rst:55]: Hmm, I thought the return value of dynamic generator task is also an `ObjectRef` that we can get the generator by calling `ray.get()`

File: doc/source/ray-core/tasks/generators.rst
Comment (jjyao) [inline doc/source/ray-core/tasks/generators.rst:2]: Actually, is actor task supported?

File: doc/source/ray-core/tasks/generators.rst
Comment (stephanie-wang) [reply doc/source/ray-core/tasks/generators.rst:55]: This was a change that I made to try to make the semantics closer to the usual list of ObjectRefs.

The main thing that might be a bit weird is that there's an implicit ray.get when you try to access the iterator. But it's fine to pass it around to other tasks before then.

File: doc/source/ray-core/tasks/generators.rst
Comment (stephanie-wang) [reply doc/source/ray-core/tasks/generators.rst:19]: Yes, good point.

File: doc/source/ray-core/tasks/generators.rst
Comment (stephanie-wang) [reply doc/source/ray-core/tasks/generators.rst:2]: Ah yes, will add it!

File: doc/source/ray-core/tasks/generators.rst
Comment (stephanie-wang) [reply doc/source/ray-core/tasks/generators.rst:55]: I preferred this syntax over returning an `ObjectRef` directly, though, since it makes it explicit that a `num_returns="dynamic"` task does not return a normal `ObjectRef`.

Comment (ericl) [reply doc/source/ray-core/tasks/generators.rst:55]: Isn't this limiting though? For example, if you wanted to wait for two generators to be completed, now you couldn't use ray.wait.

Comment (stephanie-wang) [reply doc/source/ray-core/tasks/generators.rst:55]: You could still use `ray.wait` like this as long as we support passing ObjectRefGenerators too:
```python
g1 = gen.remote()
g2 = gen.remote()
gs, _ = ray.wait([g1, g2], num_returns=2)  # Waits until the ref lists for both are ready.
```

I think the only thing that wouldn't be supported is if you wanted to block on the results of both at the same time with the list syntax:
```python
zipped = zip(g1, g2)
for pair in zipped:
  # First iteration would wait for g1 ref list, then for g2 ref list.
  pass
```

Comment (ericl) [reply doc/source/ray-core/tasks/generators.rst:55]: This has implications for other APIs as well right? (e.g., ray.kill, passing as args). It seems at the least this would change the type signature to Union[ObjectRef, ObjectRefGenerator] for many APIs.

File: python/ray/_raylet.pyx
Comment (stephanie-wang) [reply python/ray/_raylet.pyx]: Should already be caught by the generator OOM test.

File: python/ray/tests/test_generators.py
Comment (stephanie-wang) [reply python/ray/tests/test_generators.py:219]: Yeah right now it won't work if the earlier objects were stored in plasma. Added a test case and TODO to show this behavior.

File: src/ray/object_manager/ownership_based_object_directory.cc
Comment (stephanie-wang) [reply src/ray/object_manager/ownership_based_object_directory.cc:191]: It gets used in the spill location handler.

**Comment by stephanie-wang** [comment]
All comments addressed.

There is one main TODO for handling the case where a generator errors after yielding some objects in plasma. Left several notes about this in the code and tests and plan to handle in a follow-up.

File: python/ray/_raylet.pyx
Comment (jjyao) [inline python/ray/_raylet.pyx:867]: I'm a bit confused. Does `dynamic_returns == NULL` mean that we are doing reconstruction? From line 815, seems it's not.

File: python/ray/_raylet.pyx
Comment (jjyao) [reply python/ray/_raylet.pyx:2172]:
```
cdef store_task_output(
            self, serialized_object,
            const CObjectID &return_id,
            const CObjectID &generator_id,
            size_t data_size, shared_ptr[CBuffer] &metadata, const c_vector[CObjectID]
            &contained_id, int64_t *task_output_inlined_bytes,
            shared_ptr[CRayObject] *return_ptr)
```

Seems `store_task_output` doesn't accept address. Could you point me to the line of code where it's used?

File: python/ray/_raylet.pyx
Comment (jjyao) [inline python/ray/_raylet.pyx:2189]: Do we support empty dynamic generator?

File: python/ray/_raylet.pyx
Comment (jjyao) [inline python/ray/_raylet.pyx:2211]: Shouldn't it be `if` instead of `while`?

File: python/ray/_raylet.pyx
Comment (jjyao) [inline python/ray/_raylet.pyx:2175]: Are they duplicated: if outer_id is not Nil then is_dynamic is True?

File: python/ray/tests/test_generators.py
Comment (jjyao) [inline python/ray/tests/test_generators.py:153]: Do we have test for empty dynamic generator?

File: src/ray/core_worker/core_worker.h
Comment (jjyao) [inline src/ray/core_worker/core_worker.h:350]: Sometimes we are using `const std::optional<ObjectID> &generator_id` and sometimes `ObjectID::Nil()`, let's be consistent?

File: src/ray/core_worker/reference_count.cc
Comment (jjyao) [inline src/ray/core_worker/reference_count.cc:219]: Add a check to make sure?

File: python/ray/tests/test_generators.py
Comment (jjyao) [inline python/ray/tests/test_generators.py:153]: Can we test both task and actor task for dynamic generator?

**Review by jjyao — COMMENTED** [review]
lg!

File: python/ray/_raylet.pyx
Comment (stephanie-wang) [reply python/ray/_raylet.pyx:2172]: Ah yes, you're right, I'm pulling it directly from CoreWorker state now. Will remove!

File: python/ray/_private/ray_option_utils.py
Comment (stephanie-wang) [reply python/ray/_private/ray_option_utils.py:125]: I don't think the syntax here affects what syntax we use in Java. Prefer "dynamic" since it's more clear than -1.

File: python/ray/_raylet.pyx
Comment (stephanie-wang) [reply python/ray/_raylet.pyx:165]: Yes, that's possible. The TODO is an optimization for cases where we never use the inner ObjectRefs separately from the outer one. Like if you pass the generator to another task, and that task calls ray.get on each inner ref but deletes the refs before it returns. You can sort of "batch" the ref counts in that case by passing the outer ref to the task instead of the deserialized ObjectRefGenerator.

File: python/ray/_raylet.pyx
Comment (stephanie-wang) [reply python/ray/_raylet.pyx:867]: It means that the caller set `num_returns="dynamic"`.

File: python/ray/_raylet.pyx
Comment (stephanie-wang) [reply python/ray/_raylet.pyx:2189]: It should work. Will add a test case.

File: python/ray/_raylet.pyx
Comment (stephanie-wang) [reply python/ray/_raylet.pyx:2175]: Yeah I'll dedupe.

File: src/ray/core_worker/reference_count.cc
Comment (stephanie-wang) [reply src/ray/core_worker/reference_count.cc:219]: Actually I don't think the comment is accurate. Either the outer object went out of scope already without deserializing the inner one, so we don't need to add the dynamic ref. Or the outer object went out of scope after deserializing the inner one, so the dynamic ref is already added. Will update.

File: python/ray/tests/test_generators.py
Comment (stephanie-wang) [reply python/ray/tests/test_generators.py:153]: Good call, turns out it does not yet work on actors :D I'll leave it for a follow-up to unblock other work.

**Comment by jjyao** [comment]
Ping me when it's ready for more reviews :)

**Comment by stephanie-wang** [comment]
> Ping me when it's ready for more reviews :)

I think all the comments were addressed already, trying to fix CI now.

**Review by pcmoritz — APPROVED** [review]
Thanks, LGTM! Before merging, let's please mark the num_returns="dynamic" API as experimental.

Before we declare it stable, in particular I'm curious why the num_returns="dynamic" option doesn't return a generator of ObjectRefs (from the API perspective that seems more natural for me). Are there implementation limitations for this?

File: python/ray/_private/worker.py
Comment (jjyao) [inline python/ray/_private/worker.py:2269]: nit: change to a single string

File: python/ray/_private/worker.py
Comment (jjyao) [inline python/ray/_private/worker.py:2267]: `isinstance` takes 2 arguments.

File: python/ray/_raylet.pyx
Comment (jjyao) [inline python/ray/_raylet.pyx:864]: Is this comment accurate: this code can run during the first execution?

File: python/ray/_raylet.pyx
Comment (jjyao) [inline python/ray/_raylet.pyx:870]: there is no `is_dynamic` parameter now.

File: python/ray/_raylet.pyx
Comment (jjyao) [inline python/ray/_raylet.pyx:2185]: This doesn't necessarily mean it's the first execution of the task? It can also mean the generator is empty?

Are we able to catch this case: the first execution returns an empty generator but re-execution returns a non-empty generator.

File: python/ray/_raylet.pyx
Comment (jjyao) [reply python/ray/_raylet.pyx:2202]: What about this? Is it possible that this loop will be executed more than once?

File: src/ray/core_worker/core_worker.h
Comment (jjyao) [reply src/ray/core_worker/core_worker.h:350]: How do we decide when to use which style? What's the guideline we should follow in the future?

**Review by jjyao — COMMENTED** [review]
My last few comments :)

**Comment by stephanie-wang** [comment]
> Thanks, LGTM! Before merging, let's please mark the num_returns="dynamic" API as experimental.
> 
> Before we declare it stable, in particular I'm curious why the num_returns="dynamic" option doesn't return a generator of ObjectRefs (from the API perspective that seems more natural for me). Are there implementation limitations for this?

The main reason right now was to avoid complicating `ray.wait`. Here are the pros/cons for returning a generator directly:

pros:
- more consistent with the usual semantics of num_returns=x (directly return a generator of refs instead of an objectref containing a generator of refs). this would also make it simpler to swap between code that's using a static vs dynamic num_returns, since the caller code doesn't need to insert an extra ray.get
- more explicit that the function is returning a generator instead of a normal value
cons:
- accessing the generator for the first time will implicitly block until we know how many refs to return. might be a gotcha for users.
- need to extend ray.wait to support ObjectRefGenerators in addition to ObjectRefs, whereas before you could just wait on the returned ObjectRef. i don't think other APIs are affected

File: python/ray/_raylet.pyx
Comment (stephanie-wang) [reply python/ray/_raylet.pyx:2185]: Hmm good catch, let me check. We'll probably have to resolve this case as a follow-up.

File: python/ray/_raylet.pyx
Comment (stephanie-wang) [reply python/ray/_raylet.pyx:2202]: I think it's okay either way; this way is a bit more defensive to me.

File: src/ray/core_worker/core_worker.h
Comment (stephanie-wang) [reply src/ray/core_worker/core_worker.h:350]: I think either is fine and I don't feel that it's necessary to have a guideline on it. If you have a strong preference, I can change one of them to the other.

**Review by jjyao — APPROVED** [review]

**Comment by stephanie-wang** [comment]
All new tests passing.

**Pull request merged**
